README.rst
===============================
Python DNS Failover
===============================

.. image:: https://badge.fury.io/py/python-dns-failover.png
    :target: http://badge.fury.io/py/python-dns-failover
    
.. image:: https://travis-ci.org/marccerrato/python-dns-failover.png?branch=master
        :target: https://travis-ci.org/marccerrato/python-dns-failover

.. image:: https://pypip.in/d/python-dns-failover/badge.png
        :target: https://crate.io/packages/python-dns-failover?version=latest


Python script to automatically update DNS records for a bunch of servers participating in a Round-Robin DNS setup.

* Free software: BSD license
* Documentation: http://python-dns-failover.rtfd.org.

This project is mainly based on `@dotcloud`_'s work on his project `autodnsfailover`_.

The main difference is that **python-dns-failover** script is aimed to **run in an external
machine** while `autodnsfailover`_ is to run in the participating servers themselves.

.. _`@dotcloud`: https://github.com/dotcloud/
.. _`autodnsfailover`: https://github.com/dotcloud/autodnsfailover

DNS Backends
----------------
A DNS backend is responsible for updating the DNS records. It must provide the following methods:

* **get_a_records(fqdn)**: Returns the list of ip adresses records associated with the given FQDN.
* **add_a_record(fqdn, ip)**: Adds a resource record of type A to the DNS list. Optionally return the new record created.
* **delete_a_record(fqdn, ip)**: Deletes all DNS A-type resource records targeting the given ip. Optionally return the number of deleted records.

The constructor will typically be implementation-dependent, and allow to set the credentials and/or the zone to act upon.

Available DNS Backends
~~~~~~~~~~~~~~~~~~~~~~~
* `CloudFlareDNS`_
	- `email`: E-mail address of your CloudFlare account.
	- `key`: API key of your CloudFlare account.
	- `zone`: target DNS full qualified domain name.
	- `ttl`: TTL of record in seconds. 1 = Automatic, otherwise, value must in between 120 and 4,294,967,295 seconds. Defaults to 1.
	- `url`: CloudFlare client gateway interface url. Defaults to 'https://www.cloudflare.com/api_json.html'.

.. _`CloudFlareDNS`: http://www.cloudflare.com/

TODO
--------
* Documentation
* Testing

HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2013-12-29)
++++++++++++++++++

* Core functionality.
* CloudFlare DNS backend.
* First release on PyPI.

tests/test_dns_failover.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dns_failover
----------------------------------

Tests for `dns_failover` module.
"""

import unittest

# import dns_failover


class TestDNSFailover(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()


tests/test_python-dns-failover.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_python-dns-failover
----------------------------------

Tests for `python-dns-failover` module.
"""

import unittest

from python-dns-failover import python-dns-failover


class TestPython-dns-failover(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass

if __name__ == '__main__':
    unittest.main()

dns_failover/core.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import errno
import httplib
import os
import select
import signal
import time


class HttpCheck(object):
    """
    Check that a given HTTP test request can be done correctly.
    """
    def __init__(self, method='GET', url='/', body=None, headers={},
                 port=None, useHttps=False, validStatusCodes=[200, 302]):
        self.method = method
        self.url = url
        self.body = body
        self.headers = headers
        if not port:
            port = 80 if not useHttps else 443
        self.port = port
        self.useHttps = useHttps
        self.validStatusCodes = validStatusCodes

    def check(self, ipaddr):
        if self.useHttps:
            connection = httplib.HTTPSConnection(ipaddr, self.port)
        else:
            connection = httplib.HTTPConnection(ipaddr, self.port)

        try:
            connection.request(self.method, self.url, self.body, self.headers)
            response = connection.getresponse()
            return response.status in self.validStatusCodes
        except:
            return False


class TickTimer(object):
    """
    Schedule a check to be executed every *interval* seconds.
    If the check has not returned after *timeout* seconds, kill it.
    The *timeout* is fixed.
    """

    def __init__(self, interval, timeout, retry):
        self.interval = interval
        self.timeout = timeout
        self.retry = retry
        self.last = 0

    def getNextCheckTime(self):
        # Initialization case
        if self.last == 0:
            # Add one second to make sure that the first check is not
            # scheduled in the past
            self.last = time.time() + 1
            return self.last
        # Compute the timestamp of the next check
        self.last += self.interval
        # If we're already late, offset our clock
        if self.last < time.time():
            self.last = time.time()
        return self.last

    def getCheckTimeout(self):
        return self.timeout

    def getRetry(self):
        return self.retry


def boundedCheck(target, check, timer, logger):
    """
    Execute the given *check* on the given *target* (*target* should be an
    IP address). The *timer* is used to retrieve the timeout, and useful
    information will be sent using the *logger*.

    This function will fork(), and the check will be executed in the child
    process. The parent process will wait for the child process, and kill
    it if it did not answer within the timeout specified by the *timer*
    implementation.
    """
    timeout = timer.getCheckTimeout()
    deadline = time.time() + timeout
    logger.debug('[DEBUG] Starting check {0} on {1}, timeout={2}'
                 .format(check, target, timeout))
    # Use self-pipe trick: setup a SIGCHLD handler to write 1 byte to a pipe
    # (and select() on that pipe)
    pipe = os.pipe()

    def sigchld(sig, frame):
        try:
            os.write(pipe[1], ' ')
        except:
            pass

    signal.signal(signal.SIGCHLD, sigchld)
    pid = os.fork()
    if pid:
        # parent process: wait for the child
        while time.time() < deadline:
            timeout = max(0, deadline - time.time())
            try:
                rfds, wfds, efds = select.select([pipe[0]], [], [], timeout)
            except select.error as err:
                if err.args[0] == errno.EINTR:
                    continue
            if rfds:
                # something in the pipe = got a SIGCHLD
                logger.debug('[DEBUG] Child exited, retrieving its status')
                childpid, status = os.wait()
                logger.debug('[DEBUG] Child exit status={0}'.format(status))
                retval = (status == 0)
            else:
                # timeout
                logger.info('[INFO] Child timeout, killing it')
                os.kill(pid, signal.SIGKILL)
                logger.debug('[DEBUG] Reaping child process')
                os.wait()
                retval = False
            os.close(pipe[0])
            os.close(pipe[1])
            logger.debug('[DEBUG] Check result is {0}'.format(retval))
            return retval
    else:
        # child process: do the check
        try:
            if check.check(target):
                exit(0)
            else:
                exit(1)
        except Exception:
            exit(2)


def retryBoundedCheck(target, check, timer, logger):
    retry = timer.getRetry()
    failed = 0
    while True:
        if boundedCheck(target, check, timer, logger):
            # Success!
            if failed:
                # If there were previous failures, log them.
                logger.info('[INFO] Check failed {0} time(s) for target {1} '
                            'before passing.'
                            .format(failed, target))
            return True
        failed += 1
        if failed >= retry:
            logger.info('[INFO] Check failed {0} times for target {1}. '
                        'Giving up.'
                        .format(failed, target))
            return False


def run(fqdns, ip_addresses, dns, check, timer, logger):
    """
    This is the "main loop". It will repeatedly check that the machines
    pointed by the DNS records are fine.
    """
    if isinstance(fqdns, basestring):
        fqdns = [fqdns]

    if isinstance(ip_addresses, basestring):
        ip_addresses = [ip_addresses]

    logger.info('[INFO] Python-dns-failover starting')

    while True:
        now = time.time()
        nextCheck = timer.getNextCheckTime()

        # Log if checking is late
        if now > nextCheck:
            logger.warning('[WARNING] We are late by {0} seconds'
                           .format(now-nextCheck))

        # Wait until newt check time
        while time.time() < nextCheck:
            wait = nextCheck - time.time()
            logger.info('[INFO] Waiting {0} seconds before next round of '
                        'checks'
                        .format(wait))
            time.sleep(wait)

        # For each FQDN of the provided list
        for fqdn in fqdns:
            logger.debug('[DEBUG] Getting DNS records for {0}'
                         .format(fqdn))

            # Get current ip adresses records associated with the FQDN
            try:
                records = dns.get_a_records(fqdn)
            except Exception as e:
                logger.error('[ERROR] Error while retrieving DNS records for '
                             '{0}: {1}'
                             .format(fqdn, e))
                continue
            else:
                logger.debug('[DEBUG] DNS current records: {0}'
                             .format(records))

            logger.debug('[DEBUG] Checking servers')

            # For each ip address of the provided list
            for ip_addr in ip_addresses:

                # Server seems dead
                if not retryBoundedCheck(ip_addr, check, timer, logger):

                    # Server is still in the DNS
                    if ip_addr in records:

                        # Server is the last in the DNS
                        if len(records) < 2:
                            logger.error('[ERROR] Server {0} seems dead and '
                                         'it is on a DNS record for {1}, but '
                                         'it will not be removed since it is '
                                         'the last server.'
                                         .format(ip_addr, fqdn))

                        # Server is not the last. Delete it.
                        else:
                            logger.error('[ERROR] Server {0} seems dead and '
                                         'it is on a DNS record for {1}. '
                                         'Removing it.'
                                         .format(ip_addr, fqdn))
                            try:
                                n_deleted = dns.delete_a_record(fqdn, ip_addr)
                            except Exception as e:
                                logger.error('[ERROR] Error while deleting '
                                             'record {0} for {1}: {2}'
                                             .format(ip_addr, fqdn, e))
                            else:
                                logger.info('[INFO] Records targeting {0} '
                                            'were removed from the DNS ({1}).'
                                            .format(ip_addr, n_deleted))
                                # Only remove one server at a time
                                break

                    # Server not in the DNS
                    else:
                        logger.info('[INFO] Server {0} continues dead.'
                                    .format(ip_addr))

                # Server seems alive
                else:

                    # Server is already in the DNS
                    if ip_addr in records:
                        logger.info('[INFO] Server {0} seems alive and it is '
                                    'on the DNS'
                                    .format(ip_addr))

                    # Server is not in the DNS. Add it.
                    else:
                        logger.warning('[WARNING] Server {0} seems alive and '
                                       'it is not on a DNS record for {1}. '
                                       'Adding it.'
                                       .format(ip_addr, fqdn))
                        try:
                            dns.add_a_record(fqdn, ip_addr)
                        except Exception as e:
                            logger.error('[ERROR] Error while adding record '
                                         '{0} for {1}: {2}'
                                         .format(ip_addr, fqdn, e))
                        else:
                            logger.info('[INFO] Record targeting {0} was '
                                        'added to the DNS.'
                                        .format(ip_addr))


examples/example1.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import dns_failover
import logging
import logging.handlers
import os

# DNS name used by the round-robin setup
fqdn = 'sub.example.com'

# IP addresses used by the round-robin setup
ip_addresses = ['1:2:3:4', '5:6:7:8', '9:10:11:12']

# CloudFlare DNS configuration
cloudflare_email = os.getenv('CLOUDFLARE_EMAIL', 'sample@example.com')
cloudflare_key = os.getenv('CLOUDFLARE_API_KEY', '8afbe6dea02407989af4dd4c97bb6e25')
cloudflare_zone = 'example.com'

# Implementation to be used to update the DNS records
dns = dns_failover.backends.CloudFlareDNS(
    email=cloudflare_email,
    key=cloudflare_key,
    zone=cloudflare_zone,
)

# Implementation to be used to check if a server is alive
check = dns_failover.HttpCheck()

# Implementation to be used for timing parameters (how often to run the checks)
timer = dns_failover.TickTimer(interval=300, timeout=3, retry=5)

# Logger to use
logger = logging.getLogger('dns_failover')
logger.setLevel(logging.DEBUG)

# Add console handler to tollger
console_handler = logging.StreamHandler()
console_handler.setLevel(logging.INFO)
logger.addHandler(console_handler)

# Email configuration w/ Gmail
email_from_addr = os.getenv('EMAIL_FROM_ADDR', 'DNS Failover <from@example.com>')
email_to_addr = os.getenv('EMAIL_TO_ADDR', ['to1@example.com', 'to2@example.com'])
email_host = os.getenv('EMAIL_HOST', 'smtp.gmail.com')
email_port = os.getenv('EMAIL_PORT', 587)
email_host_user = os.getenv('EMAIL_HOST_USER', 'from@example.com')
email_host_password = os.getenv('EMAIL_HOST_PASSWORD', 'Password123456')
email_use_tls = os.getenv('EMAIL_USE_TLS', True)
email_subject = os.getenv('EMAIL_SUBJECT', "DNS Failover Log")

# Add email handler to logger
email_handler = logging.handlers.SMTPHandler(
    mailhost=(email_host, email_port),
    fromaddr=email_from_addr,
    toaddrs=email_to_addr,
    subject=email_subject,
    credentials=(email_host_user, email_host_password),
    secure=() if email_use_tls else None,
)
email_handler.setLevel(logging.WARNING)
logger.addHandler(email_handler)

# Run the DNS updater with all the previously set parameters!
dns_failover.run(fqdn, ip_addresses, dns, check, timer, logger)


dns_failover/backends.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import requests


class CloudFlareDNS(object):
    """
    CloudFlare Backend for Python DNS Failover
    """

    def __init__(self, email, key, zone, ttl=1,
                 url='https://www.cloudflare.com/api_json.html'):
        """
        Sets up a CloudFlareDNS backend instance with the provided
        configuration.

        # Params:

        - `email`: E-mail address of your CloudFlare account.
        - `key`: API key of your CloudFlare account.
        - `zone`: target DNS full qualified domain name.
        - `ttl`: TTL of record in seconds. 1 = Automatic, otherwise, value
                 must in between 120 and 4,294,967,295 seconds. Defaults to 1.
        - `url`: CloudFlare client gateway interface url. Defaults to
                 'https://www.cloudflare.com/api_json.html'.
        """
        self.url = url
        self.zone = zone
        self.ttl = ttl

        self.base_data = {
            'email': email,
            'tkn': key,
            'z': self.zone,
        }

    def _do_request(self, data={}):
        """
        Configures and does the request to the backend API endpoint
        and catches any possible exception.

        # Params:

        - `data`: additional data for the request.
        """
        data.update(self.base_data)

        response = requests.post(self.url, data=data)
        response.raise_for_status()
        response_data = response.json()

        # If result is not successful, raise error
        if response_data.get('result') != 'success':
            raise Exception(response_data.get('msg'))

        return response_data

    def _hostname(self, fqdn):
        """
        Asserts that the given FQDN belong to the configured zone and
        returns the hostname.

        # Params:

        - `fqdn`: full qualified domain name to retrieve the hostname from.
        """
        zone = '.' + self.zone
        assert fqdn.endswith(zone)
        return fqdn[:-len(zone)]

    @property
    def _records(self):
        """
        Load all current DNS zone records.
        Returns the list of the current DNS zone records.
        """
        data = {
            'a': 'rec_load_all',
        }
        response = self._do_request(data=data)
        return response.get('response').get('recs').get('objs')

    def get_a_records(self, fqdn):
        """
        Returns the list of ip adresses records associated with the given FQDN.

        # Params:

        - `fqdn`: full qualified domain name of the records to retrieve.
        """
        return [record.get('content')
                for record in self._records
                if record.get('name') == fqdn and record.get('type') == 'A']

    def add_a_record(self, fqdn, ip):
        """
        Adds a resource record of type A to the DNS list.
        Returns the new record created.

        # Params:

        - `fqdn`: full qualified domain name to add to the dns record.
        - `ip`: server ip to add to the dns record.
        """

        data = {
            'a': 'rec_new',
            'content': ip,
            'name': self._hostname(fqdn),
            'ttl': self.ttl,
            'type': 'A',
        }
        response = self._do_request(data=data)
        new_record = response.get('response').get('rec').get('obj')

        return new_record

    def delete_a_record(self, fqdn, ip):
        """
        Deletes all DNS A-type resource records targeting the given ip.
        Returns the number of deleted records.

        # Params:

        - `fqdn`: full qualified domain name of the dns record to remove.
        - `ip`: target ip to remove from the dns record to remove.
        """
        num_deleted = 0
        for record in self._records:
            if record.get('name') == fqdn and \
               record.get('type') == 'A' and \
               record.get('content') == ip:
                data = {
                    'a': 'rec_delete',
                    'id': record.get('rec_id'),
                }
                self._do_request(data=data)
                num_deleted += 1

        return num_deleted
